src/dot-protocol/merlin_dot_protocol.ml

```diff
@@ -127,39 +127,94 @@ module Sexp = struct
     List (List.map ~f directives)
 end
 
-module Csexp = Csexp.Make (Sexp)
+type read_error =
+  | Unexpected_output of string
+  | Csexp_parse_error of string
+
+type command = File of string | Halt | Unknown
+
+module type S = sig
+  type 'a io
+  type in_chan
+  type out_chan
 
-module Commands = struct
-  type t = File of string | Halt | Unknown
+  (** [read] reads one csexp from the channel and returns the list of
+      directives it represents *)
+  val read :
+    in_chan -> (directive list, read_error) Merlin_utils.Std.Result.t io
 
-  let read_input in_channel =
-    let open Sexp in
-    match Csexp.input in_channel with
-    | Ok (List [Atom "File"; Atom path]) -> File path
-    | Ok (Atom "Halt") -> Halt
-    | Ok _ -> Unknown
-    | Error _msg -> Halt
+  val write : out_chan -> directive list -> unit io
 
-  let send_file ~out_channel path =
-    Sexp.(List [Atom "File"; Atom path])
-    |> Csexp.to_channel out_channel
+  module Commands : sig
+    val read_input : in_chan -> command io
+
+    val send_file : out_chan -> string -> unit io
+
+    val halt : out_chan -> unit io
+  end
 end
 
-type read_error =
-  | Unexpected_output of string
-  | Csexp_parse_error of string
+module Make (IO : sig
+  type 'a t
 
-let read ~in_channel =
-  match Csexp.input in_channel with
-  | Ok (Sexp.List directives) ->
-      Ok (List.map directives ~f:Sexp.to_directive)
-  | Ok sexp ->
-    let msg = Printf.sprintf
-      "A list of directives was expected, instead got: \"%s\""
-      (Sexp.to_string sexp)
-    in
-    Error (Unexpected_output msg)
-  | Error msg -> Error (Csexp_parse_error msg)
+  module O : sig
+    val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+  end
+end) (Chan : sig
+  type in_chan
+  type out_chan
+
+  val read : in_chan -> (Csexp.t, string) result IO.t
+
+  val write : out_chan -> Csexp.t -> unit IO.t
```

Comment:
I'm not sure. I think the fork uses a single csexp as well. 
https://github.com/ocaml/merlin/blob/209e5f24f9a3a98d4319717c7b1738eb78d2e175/src/dot-protocol/merlin_dot_protocol.mli#L105

Reply:
@ddickstein you might be using an old version somewhere.

Reply:
I think you might be misreading. The functor takes a module whose `write` is implemented in terms of a single csexp. The module that is provided, `Lev_fiber_csexp.Session`, implements `write` in terms of a list, so there's a redefinition of `write` that implements it for the functor by packaging the csexp into a singleton list and calling `Lev_fiber_csexp.Session.write`. The incompatibility I was pointing out as an issue was on the read side, where the latest version of Merlin now expects the implementation that returns a result, but `Lev_fiber_csexp.Session.read` returns an option, and there's no re-implementation to satisfy the functor because ocaml-lsp hasn't yet updated to the latest version of Merlin. But I am now on latest Merlin and on latest ocaml-lsp, so I'm seeing an incompatibility.

Reply:
Yes, not sure why the `result` turned into an option, we can easily change that.

However even with that change it is important to note that the current Merlin master should not be expected to be compatible with ocaml-lsp. We still have an upstreaming PR to merge (#1585) and then we will have to make changes to ocaml-lsp to actually use the new api. Only then both projects will we be compatible for the first time.

Reply:
Implementation used internally by merlin ("Blocking"), uses Csexp library. `read` is implemented as:
```ocaml
let read = Csexp.input
```
which has type of
```
val input : in_channel -> (Sexp.t, string) result
```

Reply:
Whoops, I should have taken a closer look.

> I decided to use result, and then adjust implementation in ocaml-lsp

How does one know that we've read until EOF in this current implementation?

Reply:
Whoops, sorry for not reading this closer.

> I decided to use result, and then adjust implementation in ocaml-lsp

What would be the purpose of the string in Error case?

Reply:
Currently, the Error string is not used. I reused the old implementation, which worked that way. Do you think it would be a better idea to change the interface to use option?

Reply:
Nah, the result type is fine. I'm just curious what you plan to use it for. If it's an error message, It's not really possible to display it in a friendly way to an lsp user through the merlin API.